javalib/src/main/scala/java/time/Duration.scala

```diff
@@ -0,0 +1,232 @@
+
+  def multipliedBy(multiplicand: Long): Duration =
+    Duration(totalNanos * multiplicand)
+
+  def dividedBy(divisor: Long): Duration = Duration(totalNanos / divisor)
+
+  def negated(): Duration = Duration(totalNanos * -1)
+
+  def abs(): Duration = if (isNegative()) negated() else this
+
+  def addTo(temporal: Temporal): Temporal =
+    temporal.plus(seconds, ChronoUnit.SECONDS).plus(nanos, ChronoUnit.NANOS)
+
+  def subtractFrom(temporal: Temporal): Temporal =
+    temporal.minus(seconds, ChronoUnit.SECONDS).minus(nanos, ChronoUnit.NANOS)
+
+  def toDays(): Long = seconds / 86400
+
+  def toHours(): Long = seconds / 3600
+
+  def toMinutes(): Long = seconds / 60
+
+  def toMillis(): Long = {
+    val (quot, rem) = totalNanos /% 1000000
+    val millis = if (rem < 0) quot - 1 else quot
+    // Reference implementation throws if millis < -9223372036854775000!
+    if (millis >= -9223372036854775000L && millis <= Long.MaxValue)
+      millis.toLong
+    else
+      throw new ArithmeticException("Long overflow")
+  }
+
+  def toNanos(): Long = {
+    // Reference implementation throws if nanos < -9223372036000000000!
+    if (totalNanos >= -9223372036000000000L && totalNanos <= Long.MaxValue)
+      totalNanos.toLong
+    else
+      throw new ArithmeticException("Long overflow")
+  }
+
+  def compareTo(that: Duration) = totalNanos.compare(that.totalNanos)
+
+  override def toString(): String = {
+    val nanos1 = if (seconds >= 0 || nanos == 0) nanos else 1000000000 - nanos
+    val secs = if (seconds >= 0 || nanos == 0) seconds else seconds + 1
+    val mins = secs / 60
+    val secsOfMin = secs % 60
+    val hours = mins / 60
+    val minsOfHour = mins % 60
+    val hourPart = if (hours == 0) "" else hours.toString + "H"
+    val minPart = if (minsOfHour == 0) "" else minsOfHour.toString + "M"
+    val decimals = f"$nanos1%09d".reverse.dropWhile(_ == '0').reverse
+    val decimalPart = if (decimals.isEmpty) "" else "." + decimals
+    val secsPart = secsOfMin match {
+      case 0 if seconds != 0 && nanos == 0 => ""
+      case 0 if seconds < 0                => "-0" + decimalPart + "S"
+      case n                               => n.toString + decimalPart + "S"
+    }
+    "PT" + hourPart + minPart + secsPart
+  }
+
+  override def equals(that: Any): Boolean = that match {
+    case that: Duration =>
+      seconds == that.getSeconds && nanos == that.getNano
+
+    case _ => false
+  }
+
+  override def hashCode(): Int = 31 * seconds.hashCode + nanos
+}
+
+object Duration {
+  private def apply(nanos: BigInt): Duration = {
+    val (quot, rem) = nanos /% 1000000000
+    val seconds = if (rem < 0) quot - 1 else quot
+    val nanosOfSeconds = if (rem < 0) rem + 1000000000 else rem
+    if (seconds.isValidLong) new Duration(seconds.toLong, nanosOfSeconds.toInt)
+    else throw new ArithmeticException("Long overflow")
+  }
+
+  val ZERO = new Duration(0, 0)
+
+  def ofDays(days: Long): Duration = Duration(BigInt(days) * 86400000000000L)
```

Comment:
It might be a good idea to use private functions to represent `86400000000000L` in a more explicit way and let the compiler inline and constant fold them. I see that this could become a potential source of bugs as we could easily miss some 0s. For example `NANOS_IN_SECONDS * SECONDS_IN_MINUTE * MINUTES_IN_HOUR * HOURS_IN_DAY`.